### ISSUE TITLE
executable_path does not use PATH env variable?

### ISSUE THREAD

**Issue by asparc** [bitst0rm-pub/Formatter#45]
Hi!
Thanks for making such a generic formatting plugin!

I'm trying to set up `Formatter` in a generic way (same config for both Linux&Windows) for Autopep8 (as my first attempt).

In my settings, I have (abbreviated)
```
{
    "formatters": {
        "autopep8": {
            [...]
            "disable": false,
            "executable_path": "autopep8",
            [...]
        }
    }
}
```

When I run formatter in debug mode, I get:
```
[Formatter](common.py#L745): [INFO] System environments:
{
  [...]
  "PATH": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin",
  [...]
}
[Formatter](common.py#L299): [ERROR] File executable does not exist: autopep8
```

I can fix this problem by changing the settings to specify the full path of `autopep8`, like so:
```
{
    "formatters": {
        "autopep8": {
            [...]
            "disable": false,
            "executable_path": "/usr/local/bin/autopep8",
            [...]
        }
    }
}
```

However, I don't want a full path in my settings, because that path is different depending on my (operating) system.
Isn't `Formatter` supposed to search the `PATH` environment variable for `autopep8`?
Why isn't it enough to just set `"executable_path": "autopep8"`?

**Comment by bitst0rm** [comment]
Hi @asparc 

Formatter has an plan B in this scenario.

If "executable_path" is set:
- "executable_path": "/path/to/bin/autopep8" or
- "executable_path": "autopep8" <- this is the same as setting custom path to "/neverland/autopep8", which will never work.

If "executable_path" is `omitted` or set to` null`:
Then the auto detection mechanism is activated to search in PATH and other environments.

Please omit "executable_path" or `"executable_path": ""` if your autopep8 is already global as your case.

I will see to implement an plan C to just set something like: "executable_path": "autopep8"

**Comment by bitst0rm** [comment]
@asparc  as said, the plan C is implemented with the following changes:
- basename only to search on PATH, e.g: "executable_path": "autopep8"
- `interpreter_path` and `executable_path` can be both string or list now: "executable_path": ["autopep8win", "autopep8"]
- added additional info to settings:
```
            // Path to the third-party plugin executable to process formatting.
            // This option can be either a string or a list of executable paths.
            // - If this option is omitted or set to null, then the global executable
            //   on PATH will be used, if found.
            // - If this option is exactly the basename, then it will be used as the
            //   executable name and searched for on the PATH.
```
I closed this for now, please just feel free to re-open it.

**Comment by asparc** [comment]
Hi @bitst0rm,

Ah, I did not think of simply setting `"executable_path": ""`. That indeed works like a charm :-)
The new changes that you propose as a plan C are also interesting for when I want to go to more advanced configurations. 
Thank you!!

**Comment by asparc** [comment]
Hi @bitst0rm,

I actually only now found time to also configure Formatter for Windows, and I'm not getting it to work with your new settings.
In the settings, I set the autopep8 executable path to: `"executable_path": ["~/.conda/envs/formatter/Scripts/autopep8.exe"],`

Then I get the following error:

```
▋[Formatter](common.py#L917): [INFO] Current mode: User Settings
▋[Formatter](formatter.py#L23): [DEBUG] Target: C:\Users\gruijthuijsenc\Drive\Desktop\tmp\tmp2.py
▋[Formatter](formatter.py#L24): [DEBUG] Scope: source.python meta.path.python meta.generic-name.python 
▋[Formatter](formatter.py#L25): [DEBUG] Syntax: python
▋[Formatter](formatter.py#L26): [DEBUG] UID: autopep8 (method: module)
▋[Formatter](common.py#L345): [DEBUG] Global interpreter found: C:\Python\3.9-64\python.EXE
▋[Formatter](common.py#L340): [DEBUG] User executable found: C:\Users\gruijthuijsenc\.conda\envs\formatter\Scripts\autopep8.exe
▋[Formatter](common.py#L451): [DEBUG] Config [default]: C:\Users\gruijthuijsenc\AppData\Roaming\Sublime Text\Packages\User\formatter.assets\config\autopep8_rc.cfg
▋[Formatter](formatter_autopep8.py#L45): [DEBUG] Current arguments: ['C:\\Python\\3.9-64\\python.EXE', 'C:\\Users\\gruijthuijsenc\\.conda\\envs\\formatter\\Scripts\\autopep8.exe', '--global-config', 'C:\\Users\\gruijthuijsenc\\AppData\\Roaming\\Sublime Text\\Packages\\User\\formatter.assets\\config\\autopep8_rc.cfg', '-']
▋[Formatter](common.py#L573): [STATUS] File not formatted due to an error (exitcode=1):
```

Why is it calling both the interpreter and the executable together (in the `Current arguments`)? Just the executable is all it takes, no?

**Comment by bitst0rm** [comment]
Hi @asparc 

The log is correct as i see:
- Global interpreter found, and
- User executable found

I just wonder, how `autopep8.exe` could run standalone without being in a python environment, means without a python interpreter. In my system it is not possible, so `autopep8.exe` has been configured to use an interpreter automatically. I can change this behaviour, but need your help.
In your terminal, you try to see what is happening with both command:

```sh
C:\\Python\\3.9-64\\python.EXE C:\\Users\\gruijthuijsenc\\.conda\\envs\\formatter\\Scripts\\autopep8.exe --help
```

and 

```sh
C:\\Users\\gruijthuijsenc\\.conda\\envs\\formatter\\Scripts\\autopep8.exe --help
```


Formatter has an Plan B to use the generic method, but then you must design the bridge yourself:
Here, without using interpreter:
1. ~~Deactivate/Remove the `module method` with the uid `autopep8`~~
  or better, edit `autopep8` to `autopep88` and add the 4 keys: `"name"`, `"type"`, `"success_code"`,  `"args"`
3. Switch to use the `generic method`:

```js
        "autopep88": {  // autopep8 already exists, so pls chose another uid
            "name": "My Autopep8",
            "type": "beautifier",
            "success_code": 0,
            "args": [{{e}}", "--global-config", "{{c}}", "-"],

            // The rest is the same, so no need to modify
            "info": "https://github.com/hhatto/autopep8",
            "disable": false,
            "format_on_save": false,
            "format_on_paste": false,
            "new_file_on_format": false,
            "recursive_folder_format": {
                "enable": false,
                "exclude_folders_regex": [],
                "exclude_files_regex": [],
                "exclude_extensions": [],
                "exclude_syntaxes": []
            },
            "syntaxes": ["python"],
            "executable_path": ["~/.conda/envs/formatter/Scripts/autopep8.exe"], // or ["autopep8"] for cross-plattform on PATH
            "config_path": {
                "default": "${packages}/User/formatter.assets/config/autopep8_rc.cfg"
            }
            /* requires "environ": {"PYTHONPATH": ["/lib/python3.7/site-packages"]} */
        },
```
3. Restart ST.

Both methods are exactly the same. The only difference is: `I design` or `You design` stuff.
Not everything can be done with the generic method.

Update: ok, i see it does not need interpreter in conda windows:
https://github.com/bitst0rm-pub/conda_asparc/actions/runs/8374029347/job/22928359573#step:8:2
i will see to patch it soon.
I wonder a lot of people are using Formatter daily, but only you are bother to report bug.

**Comment by bitst0rm** [comment]
@asparc 
Update: I've released version 1.3.3 with this fix.
This could be the latest bug at the Formatter level, unless there's something else you discover.
Thanks for your reports.
https://github.com/bitst0rm-pub/Formatter/blob/master/messages/1.3.3.txt

**Comment by asparc** [comment]
Hi @bitst0rm,
Thanks for your super fast responses! It does work on windows for me now! Thanks a lot!
No worries about the few bugs, I totally get that it is hard to test everything with a package that supports so many features across so many languages and also different operating systems.
Cheers!
